substrate/frame/safe-mode/src/lib.rs

```diff
@@ -15,6 +15,48 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+//! # Safe Mode
+//!
+//! Trigger for stopping all extrinsics outside of a specific whitelist.
+//!
+//! ## Pallet API
+//!
+//! See the [`pallet`] module for more information about the interfaces this pallet exposes,
+//! including its configuration trait, dispatchables, storage items, events, and errors.
+//!
+//! ## Overview
+//!
+//! Safe mode is entered via two paths (deposit or forced) until a set block number.
+//! The mode is exited when the block number is reached or a call to one of the exit extrinsics is
+//! made. A `WhitelistedCalls` configuration contains all calls that can be executed in while in
+//! safe mode.
+//!
+//! ### Primary Features
+//!
+//! - Entering safe mode can be permissionless with deposit, origin based, or some combination of
+//!   those.
+//! - Separate origin configuration for entering and exiting safe mode.
+//! - Safe mode must always be configured to exit at a specific block number.
+//! - Safe mode exit block number may be extended by additional calls.
+//!
+//! ### Example
+//!
+//! Configuration of call filters:
+//!
+//! ```ignore
+//! impl frame_system::Config for Runtime {
+//!   // …
+//!   type BaseCallFilter = InsideBoth<DefaultFilter, SafeMode>;
+//!   // …
+//! }
+//! ```
+//!
+//! ## Low Level / Implementation Details
+//!
+//! ### Use Cost
+//! A storage value (`EnteredUntil`) is used to store the block safe mode will be exited on.
+//! Using the call filter will require a db read of that storage on each extrinsic.
```

Comment:
Yes and please note that the DB read will only hit the DB on the first time, after that it will be on the overlay and read quickly.

Reply:
Yea, a `StorageMap` is cached per key. So if the same Call gets sent twice, then it is cached (but not in general).  
On the `SafeMode` it is easier since it uses a `StorageValue` that gets cached from the first access on.